src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,551 @@
+# coding=utf-8
+# Copyright 2018 DPR Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch DPR model for Open Domain Question Answering."""
+
+
+import collections
+import logging
+from typing import List, Optional, Tuple
+
+import torch
+from torch import Tensor as T
+from torch import nn
+from torch.serialization import default_restore_location
+
+from .configuration_dpr import DprConfig
+from .file_utils import add_start_docstrings
+from .modeling_bert import BertConfig, BertModel
+from .modeling_utils import PreTrainedModel
+
+
+logger = logging.getLogger(__name__)
+
+DPR_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "dpr-base-uncased",
+]
+
+
+#############
+# BertEncoder
+#############
+
+
+class DprBertEncoder(BertModel):
+    def __init__(self, config, project_dim: int = 0):
+        BertModel.__init__(self, config)
+        assert config.hidden_size > 0, "Encoder hidden_size can't be zero"
+        self.encode_proj = nn.Linear(config.hidden_size, project_dim) if project_dim != 0 else None
+        self.init_weights()
+
+    @classmethod
+    def init_encoder(cls, cfg_name: str, projection_dim: int = 0, dropout: float = 0.0, **kwargs) -> "DprBertEncoder":
+        cfg = BertConfig.from_pretrained(cfg_name)
+        if dropout != 0:
+            cfg.attention_probs_dropout_prob = dropout
+            cfg.hidden_dropout_prob = dropout
+        return cls(cfg, project_dim=projection_dim)
+
+    def forward(self, input_ids: T, token_type_ids: Optional[T], attention_mask: Optional[T]) -> Tuple[T, ...]:
+        if self.config.output_hidden_states:
+            sequence_output, pooled_output, hidden_states = super().forward(
```

Comment:
don't like to wrap the forward function around a `super().forward()` call

Reply:
this can easily be replaced by bert_encodec(...) IMO.

Reply:
LOL I think this is lovely except that I would write:

```python
outputs = super.forward(
                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
            )
```

then unpack them depending output_hidden_states. 
Also, IDK when you branched off but `output_hidden_states` is now a kwarg I think.

Reply:
Ok nice. I removed the logic with `self.config.output_hidden_states`
I'll keep the inheritance I think to still be compatible with the format of the weights of the official repo